### PR TITLE
Supply first Docker and Kubernetes examples

### PR DESCRIPTION
**Pull request by anzoman**
The changes within this PR add the very first Docker and Kubernetes examples.

Depends on #7 and has to be merged afterwars.

### CODE REVIEW

**Comment by anzoman** [comment]
#7 was merged and this PR is now ready for a review.

File: kubernetes/rancher/README.md [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+machine. The deployed Kubernetes dashboard is the available on `localhost:80` and `localhost:443`. Additionally one 
+[hello-world](https://hub.docker.com/_/hello-world) Docker container is deployed in the Rancher network and it will 
+show up in the dashboard.
+
+There are 2 main Ansible playbooks separated into folders. Usage is explained in the table below:
+
+|    Role      |   Purpose    |
+|:-------------|:-------------|
+| **rancher** | Installs Kubernetes by running Rancher Docker container |
+| **hello-world** | Runs hello-world |
```
Comment (sstanovnik) [inline kubernetes/rancher/README.md:34]: I'd name this `rancher-hello-world`.

File: kubernetes/rancher/README.md [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+later passed to xOpera at the beginning of the orchestration. The inputs are explained in a table below.
+
+| Input | Description | Example
+|:-------------|:-------------|:-------------|
+| `host_ip` | IP address of your target (virtual) machine | localhost |
+
+### Running with xOpera
+When running Ansible playbooks with xOpera on a VM you should specify the username that will be used to connect to
+the virtual machine by ssh. This can be done by setting the environment variable `OPERA_SSH_USER`.
+You can also skip `OPERA_SSH_USER` when running xOpera by exporting this variable using: `export OPERA_SSH_USER=root`.
```
Comment (sstanovnik) [inline kubernetes/rancher/README.md:46]: Is this relevant to the example?

File: kubernetes/rancher/README.md (hunk 1 above)
Comment (sstanovnik) [reply kubernetes/rancher/README.md:34]: It's already named `docker-hello-world` though, so a change is needed eiter way :)

File: kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    ansible_python_interpreter: /usr/bin/python3
+  tasks:
+    - name: Run hello-world Docker container and join it to the Rancher network
+      docker_container:
+        name: hello-world
+        image: hello-world:linux
+        pull: true
+        state: started
+        labels:
+          io.rancher.container.network: "true"
```
Comment (sstanovnik) [inline kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml:13]: I don't think this does what you want it to do, does it? Running things in kubernetes means also configuring pods, services, ingresses and other seemingly unnecessary things, and this does none of it. I'd expect you to use a kubectl command to launch a service or something here, definitely not a Docker command.

File: validate.sh [hunk 4]
```diff
@@ -110,6 +110,11 @@ validate_example "cloud/openfaas/thumbnail-generator/image-resize" "service.yaml
 validate_example "cloud/platform-connection/aws-azure-connection/aws-azure" "service.yaml" "inputs.yaml"
 validate_example "cloud/platform-connection/aws-azure-connection/azure-aws" "service.yaml" "inputs.yaml"
 
+# test an example from kubernetes/docker
+validate_example "kubernetes/docker" "service.yaml" "inputs.yaml"
+# test an example from kubernetes/rancher
+validate_example "kubernetes/docker" "service.yaml" "inputs.yaml"
```
Comment (sstanovnik) [inline validate.sh:116]: The last line very likely needs to be kubernetes/rancher, right?

**Review by sstanovnik — CHANGES_REQUESTED** [review]

File: kubernetes/rancher/README.md [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+Minimum requirements for running Kubernetes with Rancher are:
+
+* Operating Systems
+  * Ubuntu 16.04 (64-bit)
+  * Red Hat Enterprise Linux 7.5 (64-bit)
+  * RancherOS 1.4 (64-bit)
+* Hardware
+  * 4 GB of Memory
+* Software
+  * Docker v1.12.6, 1.13.1, 17.03.2
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/rancher/README.md:19]: This is from: https://github.com/rancher/rancher#minimum-requirements.

I believe that these are the required versions for [Docker engine](https://docs.docker.com/engine/release-notes/17.03/). Should we reformat to be clearer?

File: kubernetes/rancher/README.md (hunk 2 above)
Comment (anzoman) [reply kubernetes/rancher/README.md:46]: `OPERA_SSH_USER` can be used to deploy for instance on a VM but I can remove that last sentence about exporting it because it's logical.

File: kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml (hunk 3 above)
Comment (anzoman) [reply kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml:13]: This was meant for Rancher to detect a running Docker container but I agree that the example would be better with some  examples using `kubectl` and `helm`. Do you have some suggestions on what to choose for the deployment?

File: validate.sh (hunk 4 above)
Comment (anzoman) [reply validate.sh:116]: True, will fix.

File: kubernetes/rancher/lib/rancher/playbooks/delete.yaml [hunk 6]
```diff
@@ -0,0 +1,10 @@
+---
+- hosts: all
+  vars:
+    ansible_python_interpreter: /usr/bin/python3
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/rancher/lib/rancher/playbooks/delete.yaml:4]: I had big some problems without this. Will try to reproduce and if everything goes ok and without errors I will remove this.

File: kubernetes/docker/lib/minio/playbooks/create.yaml [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    - name: Install prerequsiste python3 packages
+      pip:
+        name:
+          - docker==4.0.2
+          - botocore==1.15.27
+          - boto==2.49.0
+          - boto3==1.12.27
+          - minio==5.0.10
+        state: latest
+        executable: pip3
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/docker/lib/minio/playbooks/create.yaml:25]: I will try to keep just python3.

File: kubernetes/rancher/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (sstanovnik) [reply kubernetes/rancher/README.md:19]: Oh, I didn't even realize Docker changed its versioning scheme! I think it's fine like it is, I just thought the versions applied to different Docker components. Maybe just say "Docker Engine >= v1.12..............." instead?

File: kubernetes/rancher/README.md (hunk 2 above)
Comment (sstanovnik) [reply kubernetes/rancher/README.md:46]: No no, that wasn't what I meant. You don't create a VM anywhere, right? So why would you set a user for it, since everything runs on the local machine?

File: kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml (hunk 3 above)
Comment (sstanovnik) [reply kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml:13]: I think the standard thing would be to deploy a prometheus heml chart^1 using an ansible helm module^2. This should be fairly simple, since the whole point of a helm chart is to be simple.

1. https://artifacthub.io/packages/helm/prometheus-community/prometheus
2. https://docs.ansible.com/ansible/latest/collections/community/kubernetes/helm_module.html

You will need to install the community.kubernetes collection I think, since that's not included by default.

File: kubernetes/rancher/README.md (hunk 2 above)
Comment (anzoman) [reply kubernetes/rancher/README.md:46]: This is also true, well for this simple example I will skip this part. All I meant was if user had his own VM he could deploy this example onto this VM.

File: kubernetes/rancher/lib/rancher/playbooks/delete.yaml (hunk 6 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/rancher/lib/rancher/playbooks/delete.yaml]: Removed successfully.

File: kubernetes/docker/lib/docker/playbooks/create.yaml [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+
+    - name: Figure out the correct architecture (x86)
+      set_fact:
+        docker_arch: amd64
+      when: (ansible_machine is defined and (ansible_machine == 'x86' or ansible_machine == 'x86_64' or ansible_machine == 'i686' or ansible_machine == 'i386')) or (ansible_userspace_architecture is defined and (ansible_userspace_architecture == 'x86' or ansible_userspace_architecture == 'x86_64' or ansible_userspace_architecture == 'i686' or ansible_userspace_architecture == 'i386'))
+
+    - name: Figure out the correct architecture (armhf)
+      set_fact:
+        docker_arch: armhf
+      when: docker_arch is not defined
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/docker/lib/docker/playbooks/create.yaml]: Resolved.

File: kubernetes/docker/lib/minio/playbooks/create.yaml (hunk 7 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/docker/lib/minio/playbooks/create.yaml]: Removed python2.

File: kubernetes/docker/lib/minio/playbooks/create.yaml [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+          MINIO_SECRET_KEY: "{{ minio_credentials.secret_key }}"
+
+    - name: Install MinIO client (mc)
+      get_url:
+        url: https://dl.min.io/client/mc/release/linux-amd64/archive/mc.RELEASE.2019-12-24T23-41-36Z
+        dest: /root
+        mode: 'u=rwx,g=rx,o=rx'
+
+    - name: Move mc program to /bin/
+      shell: "mv mc.RELEASE.2019-12-24T23-41-36Z mc"
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/docker/lib/minio/playbooks/create.yaml]: Updated.

File: kubernetes/docker/lib/minio/playbooks/create.yaml [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+
+    - name: Move mc program to /bin/
+      shell: "mv mc /bin/"
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/docker/lib/minio/playbooks/create.yaml]: Done.

File: kubernetes/docker/lib/minio/playbooks/delete.yaml [hunk 11]
```diff
@@ -0,0 +1,20 @@
...
+  tasks:
+    - name: Remove any existing MinIO configurations
+      file:
+        state: absent
+        path: "/tmp/config.json"
+
+    - name: Remove MinIO client
+      file:
+        state: absent
+        path: "/bin/mc"
```
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/docker/lib/minio/playbooks/delete.yaml]: Done.

File: kubernetes/rancher/README.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (anzoman) [reply kubernetes/rancher/README.md]: Updated.

File: kubernetes/rancher/README.md (hunk 1 above)
Comment (anzoman) [reply kubernetes/rancher/README.md]: This was updated.

File: kubernetes/rancher/README.md (hunk 2 above)
Comment (anzoman) [reply kubernetes/rancher/README.md]: Removed.

File: validate.sh (hunk 4 above)
Comment (anzoman) [reply validate.sh]: Fixed.

File: kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml (hunk 3 above)
Comment (anzoman) [reply kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml]: Well this could be problematic as we talked. I'm able to deploy Rancher in a Docker container but for deploying Prometheus helm chart you would need a kubectl cluster config file which I wasn't able to retrieve from the Rancher container and I also believe that this is not the best practice. However, if I go to Rancher dashboard and copy kubectl config for the default `local` cluster everything works and Prometheus chart gets deployed.

Do you have some other ideas about that?

Comment (anzoman) [reply kubernetes/rancher/lib/docker-hello-world/playbooks/create.yaml]: I was able to overcome that issue by running Prometheus helm chart in a Rancher docker container with `docker exec` and everything seems to work fine.

**Review by sstanovnik — APPROVED** [review]

**Comment by sstanovnik** [comment]
Thanks!
